### PR TITLE
New Trello type for the notification hook

### PR DESCRIPTION
**Pull request by Spritekin**
## Description

Defines a new Trello notification type allowing the findings to be sent to Trello as cards. Implements issue secureCodeBox/secureCodeBox#1148. (https://github.com/secureCodeBox/secureCodeBox/issues/1148)

### Checklist

* [ x ] Test your changes as thoroughly as possible before you commit them. Preferably, automate your test by unit/integration tests.
* [ x ] Make sure `npm test` runs for the whole project. 

Sorry I was unable to run the "npm test" command as it kept throwing configuration errors. I think the problems originated here:
https://github.com/secureCodeBox/secureCodeBox/commit/3f001cf653514a6225960a6b7f29dcaf995e424c
https://github.com/secureCodeBox/secureCodeBox/runs/5836145105?check_suite_focus=true

When that change is removed the tests run fine.

However I have built the image and tested this code in a production level server and is working fine.

### CODE REVIEW

**Comment by Ilyesbdlala** [comment]
Thank you for your contribution!
In our project, we follow [DCO](https://developercertificate.org/). Each commit, as mentioned [here](https://github.com/secureCodeBox/secureCodeBox/blob/main/CONTRIBUTORS.md), has to be signed off : running the steps found [here](https://github.com/secureCodeBox/secureCodeBox/pull/1153/checks?check_run_id=6300572099) should do the trick.
You're right. `hooks/package.json` should not have the type module. Thank you for reminding me I left that piece of code in :sweat_smile:. I'll fix it.
I'll approve the PR for running workflows and put the PR on our project board so it can be reviewed.
For local testing: I'd recommend using our makefiles. We're in the process of updating our testing documentation. But FYI:
- in `securecodebox/` running `make npm-test-all` will run all unit-tests (after removing the module type from `hooks/package.json`)
- running `make test` in the corresponding module folder would run both its unit tests and integration tests.

**Comment by Ilyesbdlala** [comment]
@Spritekin I forgot to mention, as mentioned [here](https://github.com/secureCodeBox/secureCodeBox/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-and-licensing), that you also have to add yourself as a contributor to the [CONTRIBUTORS.md ](https://github.com/secureCodeBox/secureCodeBox/blob/main/CONTRIBUTORS.md)file as part of DCO :)

**Comment by Spritekin** [comment]
@Ilyesbdlala 
Added myself to contributors and all tests passed. Please review and merge at you earliest convenience.

Regards

File: hooks/notification/docs/README.ArtifactHub.md
Comment (SebieF) [inline hooks/notification/docs/README.ArtifactHub.md:352]:
```suggestion
- TRELLO_LABELS: Comma separated list of label IDs to apply to the card. If empty no labels will be applied. Default: ""
```

File: hooks/notification/docs/README.ArtifactHub.md
Comment (SebieF) [inline hooks/notification/docs/README.ArtifactHub.md:350]:
```suggestion
- TRELLO_KEY: Your Trello API key. Read the requirements above. Required.
- TRELLO_TOKEN: Your Trello API token. Read the requirements above. Required.
```

File: hooks/notification/docs/README.ArtifactHub.md
Comment (SebieF) [inline hooks/notification/docs/README.ArtifactHub.md:353]: Would be nice to list all possible configurations here: `The position of the new card. top, bottom, or a positive float`

File: hooks/notification/docs/README.ArtifactHub.md
Comment (SebieF) [inline hooks/notification/docs/README.ArtifactHub.md:403]: Unfortunately I was not able to get a connection to the trello api when using a kubernetes secret as described here. This might be a problem on my end, but maybe someone can double check? Using the API Key and Token directly in the values file worked fine, however.

File: hooks/notification/docs/README.ArtifactHub.md
Comment (SebieF) [inline hooks/notification/docs/README.ArtifactHub.md:362]:
```suggestion
# cat trello_values.yaml
```

File: hooks/notification/docs/README.ArtifactHub.md
Comment (SebieF) [inline hooks/notification/docs/README.ArtifactHub.md:404]:
```suggestion
helm upgrade --install nwh secureCodeBox/notification-hook --values trello_values.yaml
```

File: hooks/notification/hook/Notifiers/TrelloNotifier.test.ts
Comment (SebieF) [inline hooks/notification/hook/Notifiers/TrelloNotifier.test.ts:74]: It looks a little strange to me that you use "zap" here while using "Nmap" scanType above. The tests are successful at my machine, however :)

File: hooks/notification/hook/Notifiers/TrelloNotifier.test.ts
Comment (SebieF) [inline hooks/notification/hook/Notifiers/TrelloNotifier.test.ts:111]: See above (zap vs. nmap)

File: hooks/notification/hook/Notifiers/TrelloNotifier.ts
Comment (SebieF) [inline hooks/notification/hook/Notifiers/TrelloNotifier.ts:98]: Guess "ZAP" was overlooked here? I think it should be the general scanner name.

File: hooks/notification/hook/Notifiers/TrelloNotifier.ts
Comment (SebieF) [inline hooks/notification/hook/Notifiers/TrelloNotifier.ts:105]: Same as above for "zap". Running an nmap scan creates a card where Solution and Reference are undefined, which might be intended. But then it might be better to call the fields "ZAP - Solution" and "ZAP - Reference" and/or only include them if a ZAP scan was run. This could avoid some confusion and would create smaller cards without redundant information.

File: hooks/notification/hook/model/Finding.ts
Comment (SebieF) [inline hooks/notification/hook/model/Finding.ts:7]: Did you double check if this change is compatible with the other notification hooks? Since the tests are successfull I assume it is, however.

**Review by SebieF — CHANGES_REQUESTED** [review]
Very decent work :) Thanks for your contribution :+1: 

I added some suggestions below, most of them should be pretty easy and quick to fix.

**Comment by Spritekin** [comment]
Thanks @SebieF ,

Yeah first time with Zap, SecureCodeBox and JS (not my favourite language) so this was a new one for me.

Let me read and fix a bit more following your comments.

BTW, I got the notification hook working both with Slack AND trello as two channels in the same configuration, it works fine.

File: hooks/notification/docs/README.ArtifactHub.md
Comment (Spritekin) [reply hooks/notification/docs/README.ArtifactHub.md:403]: I create them with Terraform something like:

resource "kubernetes_secret" "sops_secret" {
  metadata {
    name = "trello
  }
  data = {
    key = "mytrellokey"
    token = "mytrellotoken"
  }
  type = "Opaque"
}

**Comment by Spritekin** [comment]
@Ilyesbdlala @SebieF 
I have implemented the required fixes but my workflow is blocked and needs approval. Would you please unblock it?

Regards,

File: hooks/notification/docs/README.ArtifactHub.md
Comment (Spritekin) [reply hooks/notification/docs/README.ArtifactHub.md]: Done

File: hooks/notification/docs/README.ArtifactHub.md
Comment (Spritekin) [reply hooks/notification/docs/README.ArtifactHub.md]: Done

File: hooks/notification/docs/README.ArtifactHub.md
Comment (Spritekin) [reply hooks/notification/docs/README.ArtifactHub.md]: Done

File: hooks/notification/docs/README.ArtifactHub.md
Comment (Spritekin) [reply hooks/notification/docs/README.ArtifactHub.md]: Done

File: hooks/notification/docs/README.ArtifactHub.md
Comment (Spritekin) [reply hooks/notification/docs/README.ArtifactHub.md]: Done and added reference to Card API

File: hooks/notification/docs/README.ArtifactHub.md
Comment (Spritekin) [reply hooks/notification/docs/README.ArtifactHub.md]: Done and named the other file trello_secrets.yaml

File: hooks/notification/docs/README.ArtifactHub.md
Comment (Spritekin) [reply hooks/notification/docs/README.ArtifactHub.md]: Done

File: hooks/notification/hook/Notifiers/TrelloNotifier.test.ts
Comment (Spritekin) [reply hooks/notification/hook/Notifiers/TrelloNotifier.test.ts:111]: I followed the pattern of the Slack and Slack App notifiers.

File: hooks/notification/hook/Notifiers/TrelloNotifier.ts
Comment (Spritekin) [reply hooks/notification/hook/Notifiers/TrelloNotifier.ts]: Added a prefix variable so users can customise their alert titles a bit.

File: hooks/notification/hook/Notifiers/TrelloNotifier.ts
Comment (Spritekin) [reply hooks/notification/hook/Notifiers/TrelloNotifier.ts]: Solution and Reference are now only added if the attributes map is not empty and the zap_solution and zap_solution exist in the map.

File: hooks/notification/docs/README.ArtifactHub.md
Comment (SebieF) [inline hooks/notification/docs/README.ArtifactHub.md:352]:
```suggestion
- TRELLO_LIST: Trello unique Id of the list where the cards will be placed. Required.
```
Typo still persists. Sorry to be picky here, I guess maybe something went wrong during the commit..

File: hooks/notification/hook/Notifiers/TrelloNotifier.ts
Comment (SebieF) [inline hooks/notification/hook/Notifiers/TrelloNotifier.ts:103]: Cool idea to add this customization :) Would it still be possible to include the scanner name into every finding notification? I think that this information is always useful for the user.

**Review by SebieF — CHANGES_REQUESTED** [review]
Thank you for your quick changes :) Very little left to fix, I also approved your workflow

File: hooks/notification/hook/Notifiers/TrelloNotifier.ts
Comment (SebieF) [reply hooks/notification/hook/Notifiers/TrelloNotifier.ts]: Cool idea, see my comment below

File: hooks/notification/docs/README.ArtifactHub.md
Comment (SebieF) [reply hooks/notification/docs/README.ArtifactHub.md]: Can you double check that it works with the kubectl command you provided in the Readme file? Still doesn't work for me using a secret and I am not sure why..

File: hooks/notification/hook/Notifiers/TrelloNotifier.ts
Comment (Spritekin) [reply hooks/notification/hook/Notifiers/TrelloNotifier.ts:103]: Will do

Comment (Spritekin) [reply hooks/notification/hook/Notifiers/TrelloNotifier.ts]: Done, added the scantype to the card desciption, plus formalised the return lines with a semicolon.

File: hooks/notification/docs/README.ArtifactHub.md
Comment (Spritekin) [reply hooks/notification/docs/README.ArtifactHub.md]: Fixed typo

**Comment by Spritekin** [comment]
@SebieF 
Fixed as requested. Please approve the workflow again.

File: hooks/notification/docs/README.ArtifactHub.md
Comment (SebieF) [inline hooks/notification/docs/README.ArtifactHub.md:403]:
```suggestion
stringData:
```

Here's the reason why it didn't work for me. Values in "data" must be base64 encoded, while "stringData" does that conversion automatically and allows the direct insertion of the api key and token.

Alternative: Use the following command to create the secret (might be good to include in the Readme)
` kubectl create secret generic trello-secret-test --from-literal=key=YOURAPIKEY --from-literal=token=YOURTOKEN`

**Review by SebieF — APPROVED** [review]
All good now, thanks again for your contribution! :) If you want to go further with this, it might be nice to include a douplication-check in the trello list, so that cards are not created twice for the same finding. But thats for another issue/PR.

**Review by SebieF — CHANGES_REQUESTED** [review]
See comment above.

File: hooks/notification/docs/README.ArtifactHub.md
Comment (Spritekin) [reply hooks/notification/docs/README.ArtifactHub.md]: Updated.

**Review by SebieF — APPROVED** [review]
The error was on my side: I got confused about the purpose of "Trello Labels" and put an arbitrary test value there which caused the objectID not found error. The API works fine! Thanks for your continuous efforts :)

**Pull request merged**
